### ISSUE TITLE
Support metrics using System.Diagnostics.Metrics 

### ISSUE THREAD

**Issue by jviau** [Azure/durabletask#785]
With the release of .NET 6 last year, a new metrics API was introduced. This is available in System.Diagnostics.DiagnosticSource 6.0 package, which is backwards compatible with older .net runtimes.

We should use this API to emit metrics for select DTFx scenarios. Customers can then listen to these metrics themselves and export them out of process appropriately, or use an existing SDK like OpenTelemetry to export them.

We can start with building a list of metrics we want to collect, their names, value significance, and any dimensions.

**Comment by cgillum** [comment]
One thing we've received asks for by customers is metrics for Azure Storage, like queue length. It's obviously specific to the DurableTask.AzureStorage backend but it would be useful if the different backends could add their own metrics as part of this work.

**Comment by jviau** [comment]
# Metrics

| Name | Instrument Type |  Unit | Unit ([ucum](https://ucum.org/ucum.html)) | Description |
| - | - | - | - | - |
| `dtfx.task.limit` | Counter | default unit |  `{concurrent_task_limit}` | The configured limit of concurrent tasks for this worker. Attributes will define orchestration vs activity. |
| `dtfx.task.current` | Counter | default unit | `{conccurent_task_current}` | The current concurrent tasks (activity or orchestration) running on the worker. Attributes will define orchestration vs activity. |
| `dtfx.queue.delay` | Histogram | milliseconds | `ms` | Measures the delay of the messages as they are dequeued. |
| `dtfx.errors` | Counter | default unit | `{errors}` | Number of task invocation errors. |
| `dtfx.task.duration` | Histogram | milliseconds | `{ms}` | Measures the duration of a task. |

# Attributes

| Name | Requirement level | Notes and examples |
| - | - | - |
| `dtfx.task.type` | Required | The type of task being ran. SHOULD be one of: `activity`, `orchestration` |
| `dtfx.task.name` | Required | The name of the task being ran. Example `MyOrchestration` |
| `dtfx.task.version` | Optional | The version of the task being ran. Omitted when version is `null` |

**Comment by jviau** [comment]
> One thing we've received asks for by customers is metrics for Azure Storage, like queue length. It's obviously specific to the DurableTask.AzureStorage backend but it would be useful if the different backends could add their own metrics as part of this work.

Yeah that is definitely important. But I do wonder if that is something DTFx should implement? Or should Azure Storage be responsible for that? I guess DTFx could add one for now, but have it opt-in only via some startup value.

**Comment by cgillum** [comment]
> Yeah that is definitely important. But I do wonder if that is something DTFx should implement? Or should Azure Storage be responsible for that? I guess DTFx could add one for now, but have it opt-in only via some startup value.

The problem with reporting it from DTFx is that DTFx doesn't have any concept of queues, partitions, or even work-item latency today. If we want DTFx to be able to report this, then we'll probably need to add some optional interface that the backends can implement to surface this information to DTFx.Core.

I see you added `dtfx.partition.name` and `dtfx.partition.length`. It's a little strange since not all backends have the concept of partitions (MSSQL doesn't - less sure about Service Bus). I suppose for those kinds of orchestration services, they could just report having one "default" partition, which is the full backlog size?

**Comment by jviau** [comment]
@cgillum each DTFx orchestration service can emit its own metrics. In this case, `DurableTask.AzureStorage` should be emitting those metrics under its own `Meter`. I will update my table to make that more clear.
